File: src/main/kotlin/jimin/55week/피로도.py
maxi = 0


def solution(k, dungeons):
    global maxi
    for i in range(len(dungeons)):
        if dungeons[i][0] <= k:
            visited = [False for _ in range(len(dungeons))]
            visited[i] = True
            dfs(k, dungeons, visited, i, 1)
    return maxi


def dfs(k, dungeons, visited, now, count):
    global maxi
    maxi = max(count, maxi)

    for i in range(len(dungeons)):
        if not visited[i] and dungeons[i][0] <= k - dungeons[now][1]:
            visited[i] = True
            dfs(k - dungeons[now][1], dungeons, visited, i, count + 1)
            visited[i] = False

File: src/main/kotlin/heejik/55week/피로도.kt
package heejik.`55week`

import kotlin.math.max

class 피로도 {
    var answer = 0
    var dungeonCount = 0
    var k = 0
    lateinit var dungeons: Array<IntArray>

    fun solution(_k: Int, _dungeons: Array<IntArray>): Int {
        k = _k
        dungeons = _dungeons
        dungeonCount = dungeons.size

        findDungeons(visited = emptyList(), fatigue = 0)
        return answer
    }

    fun findDungeons(visited: List<Int>, fatigue: Int) {
        if (fatigue > k) return
        for (i in dungeons.indices) {
            if (i !in visited && fatigue + dungeons[i][0] <= k) {
                findDungeons(visited.plus(i), fatigue + dungeons[i][1])
            }
        }
        answer = max(answer, visited.size)
    }
}
